tests/posix/common/src/pthread.c

```diff
@@ -992,3 +992,51 @@ ZTEST(posix_apis, test_pthread_attr_large_stacksize)
 	zassert_equal(actual_size, expect_size);
 	zassert_ok(pthread_attr_destroy(&attr));
 }
+
+static void *test_pthread_cancel_fn(void *arg)
+{
+	int oldtype = ~PTHREAD_CANCEL_DEFERRED;
+	int oldstate = ~PTHREAD_CANCEL_ENABLE;
+
+	zassert_ok(pthread_setcanceltype(PTHREAD_CANCEL_DEFERRED, &oldtype));
+	zassert_equal(oldtype, PTHREAD_CANCEL_DEFERRED);
+
+	zassert_ok(pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &oldstate));
+	zassert_equal(oldstate, PTHREAD_CANCEL_ENABLE);
+
+	/* this should be ignored */
+	pthread_testcancel();
+
+	/* this will mark it pending */
+	pthread_cancel(pthread_self());
+
+	/* assume this completes well before 500ms */
+	zassert_ok(pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, &oldstate));
```

Comment:
will this function call terminates the thread, instead of the function under test?

Reply:
No it shouldn't because of the state which is disable

Reply:
answer for myself: `pthread_setcancelstate` didn't terminate the thread because the `cancel_type` was set to `PTHREAD_CANCEL_DEFERRED` above